src/shared/XLinkData.c

```diff
@@ -225,6 +226,95 @@ XLinkError_t XLinkReadDataWithTimeout(streamId_t streamId, streamPacketDesc_t**
     return X_LINK_SUCCESS;
 }
 
+XLinkError_t XLinkReadMoveData(streamId_t streamId, streamPacketDesc_t* const packet)
+{
+    XLINK_RET_IF(packet == NULL);
+
+    float opTime = 0;
+    xLinkDesc_t *link = NULL;
+    XLINK_RET_IF(getLinkByStreamId(streamId, &link));
+    streamId = EXTRACT_STREAM_ID(streamId);
+
+    xLinkEvent_t event = {0};
+    XLINK_INIT_EVENT(event, streamId, XLINK_READ_REQ,
+                     0, NULL, link->deviceHandle);
+    event.header.flags.bitField.moveSemantic = 1;
+    XLINK_RET_IF(addEventWithPerf(&event, &opTime));
+
+    if (!event.data)
+    {
+        return X_LINK_ERROR;
+    }
+    *packet = *(streamPacketDesc_t *)event.data;
+
+    // free the allocation from movePacketFromStream()
+    // done within this same XLink module so the same C runtime is used
+    free(event.data);
+
+    if (glHandler->profEnable)
+    {
+        glHandler->profilingData.totalReadBytes += packet->length;
+        glHandler->profilingData.totalReadTime += opTime;
+    }
+
+    const XLinkError_t retVal = XLinkReleaseData(streamId);
```

Comment:
Wondering if `XLinkReleaseSpecificData` from #21 might be needed here if:
Eg:
 - Thread1 and Thread2 are blocked at waiting for the event to return
 - T1 is unblocked first, continues freeing the allocated streamPacketDesc_t but is rescheduled.
 - T2 is unblocked second, but its the first to arrive to XLinkReleaseData

Although not sure if that may affect anything in an incorrect way...

Maybe violating the "fillLevel"... If first packet was very large and second very small, then for the moment while T2 releases the incorrect packet (the big one), there is more space available as specified compared to the actual state.

Reply:
It is my understanding that for a single stream there can be only one thread operating on it. T1 and T2 can not both be reading the same stream. With that documentation/rule in place, it is guaranteed that the XLinkReadMoveData() call is operating on isolated circular queues and therefore there is no race conflict.

There are m.a.n.y problems with multiple threads operating on the same stream. And also unsafe with multiple connections/devices. These are both core flaws in XLink architecture. Intel made a queueing event pipeline but often ignores the pipeline by directly calling functions or mutating structures for streams and connections with no thread safety.

Until that architectural issues is resolved (likely in coordination with Intel) and all the code from that cascades, I do not think it is valuable to create complex hacks for move semantics in an attempt to work around the architectural flaws -- which are still there and will still cause faults even with hacks.

Instead, require/rule that there is no supported simultanous access to the same stream by multiple app threads.